source/geod24/RingBuffer.d

```diff
@@ -0,0 +1,127 @@
+module geod24.RingBuffer;
+
+import std.range;
+
+/*******************************************************************************
+
+    A ring buffer implementation with fixed size
+
+    Params:
+        T = Type of elements
+
+*******************************************************************************/
+
+class RingBuffer (T)
+{
+
+    /*******************************************************************************
+
+        Params:
+            size = Maximum number of elements
+
+    *******************************************************************************/
+
+    this (size_t size) @safe pure nothrow
+    {
+        this.storage = new T[size];
+        this.storage.length = size;
+        if (size)
+        {
+            this.write_range = cycle(this.storage[]);
+            this.read_range = cycle(this.storage[]);
+        }
+    }
+
+    /// Returns: true if full
+    bool full () @safe @nogc pure nothrow
+    {
+        return this.len == this.storage.length;
+    }
+
+    /// Returns: true if empty
+    bool empty () @safe @nogc pure nothrow
+    {
+        return this.len == 0;
+    }
+
+    /// Returns: Number of elements currently in the buffer
+    size_t length () @safe @nogc pure nothrow
+    {
+        return this.len;
+    }
+
+    /*******************************************************************************
+
+        Insert an element to the buffer
+
+        Params:
+            val = new element of type T
+
+    *******************************************************************************/
+
+    void insert () (auto ref T val) @safe @nogc pure nothrow
+    {
+        assert(!this.full());
+
+        write_range.front() = val;
+        write_range.popFront();
+        this.len++;
+    }
+
+    /// Returns: Element in the front of the buffer
+    T popFront () @safe @nogc pure nothrow
```

Comment:
This is breaking with the known definition of a range. `popFront` should be `void`. Only `front` / `back` / `opIndex` should ever return any element. It's very confusing to see client code receive something from `popFront` - it breaks assumptions for most D programmers.

If you wish to have a function which calls both `front()` and `popFront()` then I would recommend calling it something else. I wish I could come up with a good name for it.

Reply:
When I first saw that `popFront` does not return anything I was disappointed :smile: Removing and reading a value from a container should be atomic IMO.

Reply:
I understand. There's many people who think so. :) I think there is some history around it, but I'd have to dig up in the archives to find why it's designed this way.